### PR TITLE
[WIP] KAFKA-12899: Support --bootstrap-server in ReplicaVerificationTool

### PR DESCRIPTION
**Pull request by dongjinleekr**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala [hunk 1]
```diff
@@ -75,9 +75,14 @@ object ReplicaVerificationTool extends Logging {
 
   def main(args: Array[String]): Unit = {
     val parser = new OptionParser(false)
-    val brokerListOpt = parser.accepts("broker-list", "REQUIRED: The list of hostname and port of the server to connect to.")
+    val brokerListOpt = parser.accepts("broker-list", "DEPRECATED, use --bootstrap-server instead; ignored if --bootstrap-server is specified. The list of hostname and port of the server to connect to.")
                          .withRequiredArg
-                         .describedAs("hostname:port,...,hostname:port")
+                         .describedAs("HOST1:PORT1,...,HOST3:PORT3")
+                         .ofType(classOf[String])
+    val bootstrapServerOpt = parser.accepts("bootstrap-server", "REQUIRED. The list of hostname and port of the server to connect to.")
```
Comment (wenbingshen) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: Do we need to add a unit test for this new option?

**Review by wenbingshen — COMMENTED** [review]
@dongjinleekr Thanks for the pr. overall LGTM. Just a comment left.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 1 above)
Comment (dongjinleekr) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: I will. Stay tuned! :smile:

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala [hunk 2]
```diff
@@ -116,7 +121,13 @@ object ReplicaVerificationTool extends Logging {
     if (options.has(versionOpt)) {
       CommandLineUtils.printVersionAndDie()
     }
-    CommandLineUtils.checkRequiredArgs(parser, options, brokerListOpt)
+
+    val effectiveBrokerListOpt = if (options.has(bootstrapServerOpt))
+      bootstrapServerOpt
+    else
+      brokerListOpt
```
Comment (showuon) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala:129]: Is it possible that user provided both `bootstrapServerOpt` and `brokerListOpt` at the same time?

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala [hunk 3]
```diff
@@ -116,7 +121,13 @@ object ReplicaVerificationTool extends Logging {
...
+
+    CommandLineUtils.checkRequiredArgs(parser, options, effectiveBrokerListOpt)
 
     val regex = options.valueOf(topicWhiteListOpt)
     val topicWhiteListFiler = new IncludeList(regex)
```
Comment (showuon) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: typo: `topicWhiteListFiler` ->‵topicWhiteListFilter`

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 1 above)
Comment (showuon) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: add tests +1

**Review by showuon — COMMENTED** [review]
@dongjinleekr , since the KIP is not approved yet, I think the PR should mark as ‵WIP`. Left some comments. Thanks.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 2 above)
Comment (dongjinleekr) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala:129]: Yes. But `bootstrapServerOpt` has the preference. (`brokerListOpt` is deprecated, but not removed yet.)

**Comment by guozhangwang** [comment]
Looks good to me. Will wait for the unit test to be added.

**Comment by dongjinleekr** [comment]
@guozhangwang Yes, I am currently working on it. And could you kindly have a look at [the voting thread](https://lists.apache.org/thread.html/rebd427d5fd34acf5b378d7a904af2c804e7460b32d34ddbb3368776c%40%3Cdev.kafka.apache.org%3E)? As of present, +2 bindings with +2 non-bindings. One more binding is required to be passed.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 3 above)
Comment (dongjinleekr) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: Fixed.

**Comment by dongjinleekr** [comment]
Here it is. With rebasing onto the lastest trunk, I added unit tests that verify the command line parsing.

Also, for this KIP is not passed yet (+2 bindings), I prefiexed the title with '[WIP]'.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 2 above)
Comment (dongjinleekr) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala:129]: Fixed. See `ReplicaVerificationToolOptions#effectiveBrokerListOpt`.

**Comment by dongjinleekr** [comment]
@tombentley Could you kindly have a look when you are free? :bowing_man:

**Comment by dongjinleekr** [comment]
Rebased onto the latest trunk.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala [hunk 4]
```diff
@@ -73,53 +74,61 @@ object ReplicaVerificationTool extends Logging {
...
+  // Non-private for testing
+  sealed class ReplicaVerificationToolOptions(args: Array[String]) extends CommandDefaultOptions(args) {
+    private val brokerListOpt = parser.accepts("broker-list", "DEPRECATED, use --bootstrap-server instead; ignored if --bootstrap-server is specified. The list of hostname and port of the server to connect to.")
+      .withRequiredArg
+      .describedAs("HOST1:PORT1,...,HOST3:PORT3")
+      .ofType(classOf[String])
+    private val bootstrapServerOpt = parser.accepts("bootstrap-server", "REQUIRED. The list of hostname and port of the server to connect to.")
+      .requiredUnless("broker-list")
+      .withRequiredArg
+      .describedAs("HOST1:PORT1,...,HOST3:PORT3")
```
Comment (tombentley) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: `TopicCommandOtions` and `ConfigCommandOptions` describes this as `.describedAs("server to connect to")`, so I guess we should do the same here.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala [hunk 5]
```diff
@@ -73,53 +74,61 @@ object ReplicaVerificationTool extends Logging {
...
-    val maxWaitMsOpt = parser.accepts("max-wait-ms", "The max amount of time each fetch request waits.")
-                         .withRequiredArg
-                         .describedAs("ms")
-                         .ofType(classOf[java.lang.Integer])
-                         .defaultsTo(1000)
+  // Non-private for testing
+  sealed class ReplicaVerificationToolOptions(args: Array[String]) extends CommandDefaultOptions(args) {
+    private val brokerListOpt = parser.accepts("broker-list", "DEPRECATED, use --bootstrap-server instead; ignored if --bootstrap-server is specified. The list of hostname and port of the server to connect to.")
+      .withRequiredArg
+      .describedAs("HOST1:PORT1,...,HOST3:PORT3")
```
Comment (tombentley) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: See later comment as for bootstrap-server.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala [hunk 6]
```diff
@@ -73,53 +74,61 @@ object ReplicaVerificationTool extends Logging {
...
+      .withRequiredArg
+      .describedAs("bytes")
+      .ofType(classOf[java.lang.Integer])
+      .defaultsTo(ConsumerConfig.DEFAULT_MAX_PARTITION_FETCH_BYTES)
+    private val maxWaitMsOpt = parser.accepts("max-wait-ms", "The max amount of time each fetch request waits.")
+      .withRequiredArg
+      .describedAs("ms")
+      .ofType(classOf[java.lang.Integer])
+      .defaultsTo(1000)
     val topicWhiteListOpt = parser.accepts("topic-white-list", "White list of topics to verify replica consistency. Defaults to all topics.")
```
Comment (tombentley) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: If we're changing this class we ought to deprecate this option and replace it with one with a less contentious name.

File: core/src/test/scala/kafka/tools/ReplicaVerificationToolTest.scala [hunk 7]
```diff
@@ -17,14 +17,40 @@
...
+  def testConfigOptWithBootstrapServers(): Unit = {
+    val opts1 = new ReplicaVerificationToolOptions(Array("--bootstrap-server", "localhost:9092"))
+    assertEquals("localhost:9092", opts1.bootstrapServer)
+
+    val opts2 = new ReplicaVerificationToolOptions(Array("--broker-list", "127.0.0.1:9092"))
+    assertEquals("127.0.0.1:9092", opts2.bootstrapServer)
+
+    val opts3 = new ReplicaVerificationToolOptions(Array("--broker-list", "127.0.0.1:9092", "--bootstrap-server", "localhost:9092"))
+    assertEquals("localhost:9092", opts3.bootstrapServer)
+  }
```
Comment (tombentley) [inline core/src/test/scala/kafka/tools/ReplicaVerificationToolTest.scala:86]: We should have an assertion with multiple `--bootstrap-servers` given.

Comment (dongjinleekr) [reply core/src/test/scala/kafka/tools/ReplicaVerificationToolTest.scala:86]: Great. :+1:

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 4 above)
Comment (dongjinleekr) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: `ConsoleConsumer`, `AclCommand`, `ConfigCommand` also do. :+1:

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 6 above)
Comment (dongjinleekr) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: Well, `fetchSizeOpt` and `maxWaitMsOpt` are precisely the same as the previous ones. Which change do you mean? :thinking:

Comment (tombentley) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: whitelist is a term we should avoid (see [KIP-629](https://cwiki.apache.org/confluence/display/KAFKA/KIP-629%3A+Use+racially+neutral+terms+in+our+codebase) and [KAFKA-10201](https://issues.apache.org/jira/browse/KAFKA-10201)).

**Comment by dongjinleekr** [comment]
Hi @tombentley

Here is the update. I added additional tests (`ReplicaVerificationToolTest#{testExitWithMultipleBrokerLists, testExitWithMultipleBootstrapServers`), Add descriptions to exising test, with changing parameter descriptions.

**Comment by dongjinleekr** [comment]
Hi @tombentley,

Sorry for missing your comment on [KIP-629](https://cwiki.apache.org/confluence/display/KAFKA/KIP-629%3A+Use+racially+neutral+terms+in+our+codebase?src=contextnavchildmode). Reviewing [KAFKA-10201](https://issues.apache.org/jira/browse/KAFKA-10201), I found that it is already under progress with [KAFKA-10589](https://issues.apache.org/jira/browse/KAFKA-10589). So, it would be better to address this issue in [this PR](https://github.com/apache/kafka/pull/9404).

I also rebased onto the latest trunk. Thank you for reviewing! :bowing_man:

**Comment by kkonstantine** [comment]
@dongjinleekr @tombentley if we intend to include this fix to 3.0 we'll need to merge asap. Please let me know so I can update the status of https://issues.apache.org/jira/browse/KAFKA-12899 accordingly

**Comment by dongjinleekr** [comment]
@kkonstantine It would be really great for me if this PR is included in 3.0, but it still needs one more binding +1. (It is also why it is prefixed with '[WIP]'.) Could you have a look at [the voting thread](https://lists.apache.org/thread.html/rebd427d5fd34acf5b378d7a904af2c804e7460b32d34ddbb3368776c%40%3Cdev.kafka.apache.org%3E) and cast a vote? If so, it may be included in the next release.

Besides, I think we should merge [KAFKA-10589](https://github.com/apache/kafka/pull/11007) before this one; you know, [KIP-629](https://cwiki.apache.org/confluence/display/KAFKA/KIP-629:+Use+racially+neutral+terms+in+our+codebase) is already passed.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala [hunk 8]
```diff
@@ -140,25 +150,28 @@ object ReplicaVerificationTool extends Logging {
...
+    val topicIds = topicsMetadata.map(metadata => metadata.name() -> metadata.topicId()).toMap
 
     val filteredTopicMetadata = topicsMetadata.filter { topicMetaData =>
-      topicsIncludeFilter.isTopicAllowed(topicMetaData.name, excludeInternalTopics = false)
+      opts.topicsIncludeFilter.isTopicAllowed(topicMetaData.name, excludeInternalTopics = false)
     }
 
     if (filteredTopicMetadata.isEmpty) {
-      error(s"No topics found. $topicsIncludeOpt if specified, is either filtering out all topics or there is no topic.")
+      error(s"No topics found. ${opts.topicWhiteListOpt} if specified, is either filtering out all topics or there is no topic.")
```
Comment (showuon) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: Is it intended to change from `topicsIncludeOpt` to `topicWhiteListOpt` ?

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala [hunk 9]
```diff
@@ -73,55 +74,64 @@ object ReplicaVerificationTool extends Logging {
...
-                           .describedAs("timestamp/-1(latest)/-2(earliest)")
-                           .ofType(classOf[java.lang.Long])
-                           .defaultsTo(-1L)
-    val reportIntervalOpt = parser.accepts("report-interval-ms", "The reporting interval.")
-                         .withRequiredArg
-                         .describedAs("ms")
-                         .ofType(classOf[java.lang.Long])
-                         .defaultsTo(30 * 1000L)
-    val helpOpt = parser.accepts("help", "Print usage information.").forHelp()
-    val versionOpt = parser.accepts("version", "Print version information and exit.").forHelp()
```
Comment (showuon) [inline core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala:113]: Why did we remove `help` and `version` options?

**Review by showuon — COMMENTED** [review]
@dongjinleekr , thanks for the PR.
```
As of present:

Binding: +2 (Gwen, Randall)
Non-binding: +1 (Luke)
```
I think you can close this vote now since I'm also +1 binding now. :)

Left some comments. Thanks.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 8 above)
Comment (dongjinleekr) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala]: Oh no, this modification is accidentally introduced by merging; Fixed.

File: core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala (hunk 9 above)
Comment (dongjinleekr) [reply core/src/main/scala/kafka/tools/ReplicaVerificationTool.scala:113]: They are provided by `CommandDefaultOptions#(helpOpt|versionOpt)`, which `ReplicaVerificationToolOptions` extends. They are redundant' so removed.

As a note: other `CommandDefaultOptions` variants like `ConsumerConfig`, `LeaderElectionCommandOptions` don't have these opts, also.

**Comment by dongjinleekr** [comment]
@showuon Sorry for being late. Rebased onto the latest trunk, resolving the issues.

> I think you can close this vote now since I'm also +1 binding now. :)

OMG, What an unexpected present! 😆️ But could you kindly cast your vote on the [mailing thread](https://lists.apache.org/thread/rwsov8d87n5pd1wc8n13z0x2jvor932b)? As far as I know, the voting is only valid within the mailing list. :pray:

**Comment by dongjinleekr** [comment]
Closes this PR, since [KAFKA-12899](https://issues.apache.org/jira/browse/KAFKA-12899) has been replaced by [KAFKA-17073](https://issues.apache.org/jira/browse/KAFKA-17073).

Thank you to everyone who participated in the discussion. Let's continue on KAFKA-17073! :bowing_man:

**Pull request closed without merging**